### ISSUE TITLE
Transposed preconditioned GMRES

### ISSUE THREAD

**Issue by f0uriest** [jax-ml/jax#29449]
### Description

I'm using `jax.scipy.sparse.linalg.gmres` with a preconditioner to solve some linear systems, and was trying to figure out why the forward solve would converge just fine but the gradient would not. 

Looking through the code it looks like `_solve` closes over the preconditioner, so that the same preconditioner is used for the forward and backward pass:
https://github.com/jax-ml/jax/blob/e04cc283d84a2df3ab0baa9c37f19f90600e11c1/jax/_src/scipy/sparse/linalg.py#L698-L700

Assuming the preconditioner satisfies $M A \sim I$, there's no reason that $M A^T \sim I$ unless $A$ is symmetric (which it usually isn't for GMRES).

I think the correct thing to do would be to define a separate `_transpose_solve` that uses the transpose of the preconditioner, since if $M A \sim I$ then $M^T A^T \sim I$ as well.

(also note that I think this is an issue for `jax.scipy.sparse.linalg.bicgstab` as well. `cg` is fine because $A$ is symmetric)

### System info (python version, jaxlib version, accelerator, etc.)

jax:    0.6.0
jaxlib: 0.6.0
numpy:  1.26.4
python: 3.12.9 | packaged by Anaconda, Inc. | (main, Feb  6 2025, 18:56:27) [GCC 11.2.0]
device info: cpu-8, 8 local devices"
process_count: 1
platform: uname_result(system='Linux', node='Nautilus', release='6.11.0-26-generic', version='#26~24.04.1-Ubuntu SMP PREEMPT_DYNAMIC Thu Apr 17 19:20:47 UTC 2', machine='x86_64')

**Comment by shoyer** [comment]
I agree, this seems a little problematic!

> I think the correct thing to do would be to define a separate `_transpose_solve` that uses the transpose of the preconditioner, since if $M A \sim I$ then $M^T A^T \sim I$ as well.

Wouldn't it be $A^T M^T \sim I$? So $M^T$ would be a right-side preconditioner instead of a left-sdie preconditioner. Not sure exactly how that changes the numerics, but it may not be a drop in replacement.

We might need a separate argument for passing a preconditioner to transposed solve.

You might also try taking a look at Lineax to see if they solve this better:
https://github.com/patrick-kidger/lineax

**Comment by f0uriest** [comment]
> Wouldn't it be A T M T ∼ I ? So M T would be a right-side preconditioner instead of a left-sdie preconditioner. Not sure exactly how that changes the numerics, but it may not be a drop in replacement.

Strictly speaking this would be correct, but in practice I don't think you need to swap the side of the preconditioner. For `cg` the convergence behavior only depends on the eigenvalues, and $AB$ has the same spectrum as $BA$ so left and right preconditioning are the same. For `gmres` and `bicgstab` the convergence depends on both the eigenvalues and eigenvectors, so it can be slightly different for left and right preconditioned systems, but in practice there's usually little difference so which side to precondition is more of an implementation detail rather than theoretical (see eg Saad, (2003). Iterative methods for sparse linear systems). Assuming it converges, the results should be the same for either left or right preconditioned systems. (and in the limit of the "ideal preconditioner" $M = A^{-1}$ then the side doesn't matter either since $A$ always commutes with it's inverse.

I'm familiar with lineax, this bug was found when I was doing some benchmarking between the two. They were previously dropping the preconditoner entirely for the transposed solve which caused other issues, but it's recently been updated to use the transpose of the preconditioner, without swapping sides: https://github.com/patrick-kidger/lineax/pull/162

**Comment by shoyer** [comment]
OK, sounds like this would be the way to go then. A PR fixing this for `gmres` and `bicgstab` would be very welcome!
